app/charts/shared/chart-helpers.tsx

```diff
@@ -121,31 +127,166 @@ export const usePreparedData = ({
   return preparedData;
 };
 
-// Helper to pivot a dataset to a wider format (used in stacked charts)
-export const getWideData = ({
+// Helper to pivot a dataset to a wider format
+export const getBaseWideData = ({
   xKey,
   groupedMap,
   getSegment,
   getY,
 }: {
   xKey: string;
-  groupedMap: Map<string, Record<string, ObservationValue>[]>;
+  groupedMap: Map<string, Array<Observation>>;
   getSegment: (d: Observation) => string;
   getY: (d: Observation) => number | null;
-}): { [key: string]: ObservationValue }[] => {
+}): Array<Observation> => {
   const wideArray = [];
+
   for (const [key, values] of groupedMap) {
-    let obj: { [key: string]: ObservationValue } = {
+    let observation: Observation = {
       [xKey]: key,
       total: sum(values, getY),
     };
 
     for (const value of values) {
-      obj[getSegment(value)] = getY(value);
+      observation[getSegment(value)] = getY(value);
+    }
+
+    wideArray.push(observation);
+  }
+
+  return wideArray;
+};
+
+export const getTemporalWideDataWithImputedValues = ({
+  xKey,
+  groupedMap,
+  imputationType,
+  segmentKey,
+  segments,
+  getSegment,
+  getX,
+  getY,
+}: {
```

Comment:
From my very simple performance tests, it seems that drawing a chart w/o any imputations is as fast as filling with 0s, while linear interpolation is several times slower. Maybe some caching could be used there to not run this function again and again when we use the same inputs? I' would appreciate your view on that :)

Reply:
Can we find the actual slow parts (using Chrome dev tools)? I haven't looked at the code too closely yet but it seems that there are quite a few intermediate arrays created and looped over and maybe there's an opportunity to optimize this a bit.

Just by looking at the arguments of this function, memoizing seems not super-trivial, since we need to make sure none of these inputs change between renders. Not the easiest to do with React …

Reply:
Will look into that!

Reply:
It seems that we would only need

- data
- x variable
- y variable
- stack variable ("segment")
- imputation type

and then we can memoize easily. getX and xKey seem redundant and groupedMap seems to be computable through data and segmentKey right ?

I think that memoization should be done regardless of whether the underlying algorithm is performant.

I find the algorithm below hard to read, I am wondering if it could be simplified into:

```
const fillNaLinear = (values) => {
  
}

const fillNaZeros = (values) =. {}

const fillNaNone = (values) => {}
```

Here it seems like the core algorithm is obscured by our particular business logic on how to access x value, y value, segment value.